streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala

```diff
@@ -0,0 +1,110 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.scheduler
+
+import scala.collection.mutable
+import scala.util.Random
+
+import org.apache.spark.streaming.scheduler.ReceiverState._
+
+private[streaming] case class ReceiverTrackingInfo(
+    receiverId: Int,
+    state: ReceiverState,
+    scheduledLocations: Option[Seq[String]],
+    runningLocation: Option[String])
+
+private[streaming] trait ReceiverSchedulingPolicy {
+
+  /**
+   * Return a list of candidate executors to run the receiver. If the list is empty, the caller can
+   * run this receiver in arbitrary executor.
+   */
+  def scheduleReceiver(
+      receiverId: Int,
+      preferredLocation: Option[String],
+      receiverTrackingInfoMap: Map[Int, ReceiverTrackingInfo],
+      executors: Seq[String]): Seq[String]
+}
+
+/**
+ * A ReceiverScheduler trying to balance executors' load. Here is the approach to schedule executors
+ * for a receiver.
+ * <ol>
+ *   <li>
+ *     If preferredLocation is set, preferredLocation should be one of the candidate executors.
+ *   </li>
+ *   <li>
+ *     Every executor will be assigned to a weight according to the receivers running or scheduling
+ *     on it.
+ *     <ul>
+ *       <li>
+ *         If a receiver is running on an executor, it contributes 1.0 to the executor's weight.
+ *       </li>
+ *       <li>
+ *         If a receiver is scheduled to an executor but has not yet run, it contributes
+ *         `1.0 / #candidate_executors_of_this_receiver` to the executor's weight.</li>
+ *     </ul>
+ *     At last, if there are more than 3 idle executors (weight = 0), returns all idle executors.
+ *     Otherwise, we only return 3 best options according to the weights.
+ *   </li>
+ * </ol>
+ *
+ */
+private[streaming] class LoadBalanceReceiverSchedulingPolicyImpl extends ReceiverSchedulingPolicy {
```

Comment:
LoadBalanceReceiverSchedulingPolicyImpl --> LoadBalance*d*ReceiverSchedulingPolicy

Why add impl? Either use `ReceiverSchedulingPolicyImpl` or `LoadBalancedReceiverSchedulingPolicy` I am fine either ways. In face, the first one might be better since there is only one implementation